plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/precompiled/ScriptPlugin.kt

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.kotlin.dsl.plugins.precompiled
+
+import org.gradle.api.Plugin
+
+import org.jetbrains.kotlin.lexer.KotlinLexer
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.name.NameUtils
+
+import java.io.File
+
+
+internal
+data class ScriptPlugin(val scriptFile: File) {
+
+    /**
+     * Gradle plugin id inferred from the script file name and package declaration (if any).
+     */
+    val id by lazy {
+        packagePrefixed(fileNameWithoutScriptExtension)
+    }
+
+    /**
+     * Fully qualified name for the [Plugin] implementation class.
+     *
+     * The [Plugin] implementation class adapts the precompiled script class
+     * to the Gradle [Plugin] protocol and it is automatically generated by
+     * the `generateScriptPluginAdapters` task.
+     */
+    val implementationClass by lazy {
+        fileNameWithoutScriptExtension.kebabCaseToPascalCase() + "Plugin"
+    }
+
+    private
+    val fileNameWithoutScriptExtension by lazy {
+        scriptFile.name.removeSuffix(".gradle.kts")
+    }
+
+    /**
+     * Fully qualified name
+     */
+    val compiledScriptTypeName by lazy {
+        packagePrefixed(scriptClassNameForFile(scriptFile))
+    }
+
+    val packageName: String? by lazy {
+        packageNameOf(scriptFile)
+    }
+
+    private
+    fun packagePrefixed(id: String) =
+        packageName?.let { "$it.$id" } ?: id
+}
+
+
+private
+fun packageNameOf(file: File): String? =
+    packageNameOf(file.readText())
+
+
+private
+fun packageNameOf(code: String): String? =
```

Comment:
❓  given this will happen before `:kotlinCompile` I wonder about the behavior if the lexer throws.

Reply:
I would expect the build to fail.

Reply:
Right. I'm wondering *how* it would fail and if the error would make sense to the user
I think we should have a test for this and make sure we display a reasonable error.